File: javascript/prime_sieve.js
function primeSieve(n) {
    const isPrime = new Array(n + 1).fill(true);
    isPrime[0] = false;
    isPrime[1] = false;
    
    for (let i = 2; i * i <= n; i++) {
        if (isPrime[i]) {
            for (let j = i * i; j <= n; j += i) {
                isPrime[j] = false;
            }
        }
    }
    
    let count = 0;
    for (let i = 2; i <= n; i++) {
        if (isPrime[i]) {
            count++;
        }
    }
    
    return count;
}

// 兼容 Node.js 和 QuickJS
const args = typeof process !== 'undefined' ? process.argv.slice(2) : scriptArgs.slice(1);
if (args.length < 1) {
    console.log('Usage: prime_sieve <n>');
    if (typeof process !== 'undefined') process.exit(1);
    else throw new Error('Usage error');
}

const n = parseInt(args[0]);
if (isNaN(n)) {
    console.log('Invalid number:', args[0]);
    if (typeof process !== 'undefined') process.exit(1);
    else throw new Error('Usage error');
}

const start = Date.now();
const count = primeSieve(n);
const end = Date.now();

const timeMs = end - start;

console.log(`JavaScript: prime_sieve(${n}) = ${count}`);
console.log(`Time: ${timeMs}ms`);

File: python/prime_sieve.py
import sys
import time

def prime_sieve(n):
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    
    i = 2
    while i * i <= n:
        if is_prime[i]:
            j = i * i
            while j <= n:
                is_prime[j] = False
                j += i
        i += 1
    
    count = 0
    for i in range(2, n + 1):
        if is_prime[i]:
            count += 1
    
    return count

if len(sys.argv) < 2:
    print('Usage: prime_sieve <n>')
    sys.exit(1)

try:
    n = int(sys.argv[1])
except ValueError:
    print('Invalid number:', sys.argv[1])
    sys.exit(1)

start = time.time()
count = prime_sieve(n)
end = time.time()

time_ms = int((end - start) * 1000)

print(f'Python: prime_sieve({n}) = {count}')
print(f'Time: {time_ms}ms')